packages/rpc-graphql/src/loaders/program-accounts.ts

```diff
@@ -38,14 +28,62 @@ async function loadProgramAccounts(
 
 function createProgramAccountsBatchLoadFn(rpc: Rpc<GetProgramAccountsApi>) {
     const resolveProgramAccountsUsingRpc = loadProgramAccounts.bind(null, rpc);
-    return async (programAccountsQueryArgs: readonly ProgramAccountsLoaderArgs[]) =>
-        programAccountsQueryArgs.map(async args => await resolveProgramAccountsUsingRpc(applyDefaultArgs(args)));
+    return async (
+        accountQueryArgs: readonly ProgramAccountsLoaderArgs[],
+    ): Promise<Promise<ProgramAccountsLoaderValue>[]> => {
+        /**
+         * Gather all the program-accounts that need to be fetched, grouped by
+         * program address.
+         */
+        const programAccountsToFetch: ToFetchMap<ProgramAccountsLoaderArgsBase, ProgramAccountsLoaderValue> = {};
+        try {
+            return accountQueryArgs.map(
+                ({ programAddress, ...args }) =>
+                    new Promise((resolve, reject) => {
+                        const accountRecords = (programAccountsToFetch[programAddress] ||= []);
+                        accountRecords.push({ args, promiseCallback: { reject, resolve } });
+                    }),
+            );
+        } finally {
+            /**
+             * Group together program-accounts that are fetched with identical args.
+             */
+            const programAccountsFetchesByArgsHash =
+                buildCoalescedFetchesByArgsHashWithDataSlice(programAccountsToFetch);
+
+            /**
+             * For each set of program-accounts related to some common args, fetch them in the fewest
+             * number of network requests.
+             */
+            await Promise.all(
+                Object.values(programAccountsFetchesByArgsHash).map(({ args, fetches: programAddressCallbacks }) => {
+                    return Object.entries(programAddressCallbacks).map(([programAddress, { callbacks }]) => {
```

Comment:
Yeah, doubly sorry for leading you astray earlier. These, in the `finally` block, can all just be `.forEach` with no returns or awaits. You just need to kick the requests off synchronously and hand them the resolve/reject methods.